library/src/scala/Tuple.scala

```diff
@@ -82,82 +81,130 @@ sealed trait Tuple extends Product {
+  // Only bounded by `<: Tuple` not `<: NonEmptyTuple`
+  // even though it only matches non-empty tuples.
+  // Avoids bounds check failures from an irreducible type
+  // like `Tuple.Head[Tuple.Tail[X]]`
+  type Head[X <: Tuple] = X match
     case x *: _ => x
-  }
 
-  /** Type of the initial part of the tuple without its last element */
-  type Init[X <: Tuple] <: Tuple = X match {
+  /** The type of the last element of a tuple */
+  type Last[X <: Tuple] = X match
+    case x *: EmptyTuple => x
+    case _ *: xs => Last[xs]
+
+  /** The type of a tuple consisting of all elements of tuple X except the first one */
+  type Tail[X <: Tuple] <: Tuple = X match
+    case _ *: xs => xs
+
+  /** The type of the initial part of a tuple without its last element */
+  type Init[X <: Tuple] <: Tuple = X match
     case _ *: EmptyTuple => EmptyTuple
     case x *: xs =>
       x *: Init[xs]
-  }
 
-  /** Type of the tail of a tuple */
-  type Tail[X <: Tuple] <: Tuple = X match {
-    case _ *: xs => xs
+  /** The type of the tuple consisting of the first `N` elements of `X`,
+   *  or all elements if `N` exceeds `Size[X]`.
+   */
+  type Take[X <: Tuple, N <: Int] <: Tuple = N match
+    case 0 => EmptyTuple
+    case S[n1] => X match
+      case EmptyTuple => EmptyTuple
+      case x *: xs => x *: Take[xs, n1]
+
+  /** The type of the tuple consisting of all elements of `X` except the first `N` ones,
+   *  or no elements if `N` exceeds `Size[X]`.
+   */
+  type Drop[X <: Tuple, N <: Int] <: Tuple = N match {
+    case 0 => X
+    case S[n1] => X match {
+      case EmptyTuple => EmptyTuple
+      case x *: xs => Drop[xs, n1]
+    }
   }
 
-  /** Type of the last element of a tuple */
-  type Last[X <: Tuple] = X match {
-    case x *: EmptyTuple => x
-    case _ *: xs => Last[xs]
+  /** The pair type `(Take(X, N), Drop[X, N]). */
+  type Split[X <: Tuple, N <: Int] = (Take[X, N], Drop[X, N])
+
+  /** Type of a tuple with an element appended */
+  type Append[X <: Tuple, Y] <: NonEmptyTuple = X match {
+    case EmptyTuple => Y *: EmptyTuple
+    case x *: xs => x *: Append[xs, Y]
   }
 
-  /** Type of the concatenation of two tuples */
-  type Concat[X <: Tuple, +Y <: Tuple] <: Tuple = X match {
+  /** Type of the concatenation of two tuples `X` and `Y` */
+  type Concat[X <: Tuple, +Y <: Tuple] <: Tuple = X match
     case EmptyTuple => Y
     case x1 *: xs1 => x1 *: Concat[xs1, Y]
-  }
 
-  /** Type of the element at position N in the tuple X */
-  type Elem[X <: Tuple, N <: Int] = X match {
-    case x *: xs =>
-      N match {
-        case 0 => x
-        case S[n1] => Elem[xs, n1]
-      }
-  }
+  /** An infix shorthand for `Concat[X, Y]` */
+  infix type ++[X <: Tuple, +Y <: Tuple] = Concat[X, Y]
 
-  /** Literal constant Int size of a tuple */
-  type Size[X <: Tuple] <: Int = X match {
+  /** The index of `Y` in tuple `X` as a literal constant Int,
+   *  or `Size[X]` if `Y` does not occur in `X`
```

Comment:
Are we sure about this ? 
I would expect an unreducible match type, or -1, or an option

Reply:
As discussed verbally, there are good arguments to make it total and to return `Size[X]`